homeassistant/components/sensor/mychevy.py

```diff
@@ -0,0 +1,155 @@
+        """Initialize sensor with car connection."""
+        self._state = None
+        self._last_update = dt.now()
+        self._conn = connection
+        connection.status = self
+
+    def success(self):
+        """Update state, trigger updates."""
+        if self._state != MYCHEVY_SUCCESS:
+            _LOGGER.info("Successfully connected to mychevy website")
+            self._state = MYCHEVY_SUCCESS
+        self.schedule_update_ha_state()
+
+    def error(self):
+        """Update state, trigger updates."""
+        if self._state != MYCHEVY_ERROR:
+            self.hass.components.persistent_notification.create(
+                "Error:<br/>Connection to mychevy website failed. "
+                "This probably means the mychevy to OnStar link is down.",
+                title=NOTIFICATION_TITLE,
+                notification_id=NOTIFICATION_ID)
+            self._state = MYCHEVY_ERROR
+        self.schedule_update_ha_state()
+
+    @property
+    def icon(self):
+        """Return the icon."""
+        return self._icon
+
+    @property
+    def name(self):
+        """Return the name."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state."""
+        return self._state
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return False
+
+
+class EVSensor(Entity):
+    """Base EVSensor class.
+
+    The only real difference between sensors is which units and what
+    attribute from the car object they are returning. All logic can be
+    built with just setting subclass attributes.
+    """
+
+    def __init__(self, connection, config):
+        """Initialize sensor with car connection."""
+        self._conn = connection
+        connection.sensors.append(self)
+        self.car = connection.car
+        self._name = config.name
+        self._attr = config.attr
+        self._unit_of_measurement = config.unit_of_measurement
+        self._icon = config.icon
+        self._icon_func = config.icon_func
+        self._state = None
+
+        self.entity_id = ENTITY_ID_FORMAT.format(
+            '{}_{}'.format(DOMAIN, slugify(self._name)))
+
+    @property
+    def icon(self):
+        """Return the icon."""
+        if self._icon_func is not None:
+            return self._icon_func()
+        return self._icon
+
+    @property
+    def name(self):
+        """Return the name."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement the state is expressed in."""
+        return self._unit_of_measurement
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Update chevy data."""
+        # safe_update takes 2 - 3 minutes to run, it's fine to just
+        # get it started, then update with whatever we find.
+        self.hass.async_add_job(self._conn.update)
```

Comment:
If this takes 2-3 min to complete, it will occupy a worker pool thread for that time. I'm not sure if that is ok. I think we need a comment from core on this.

Reply:
Yeh, I'm not either. It is easy enough to build my own dedicated thread. I'm happy to go whatever way people would rather here.

Reply:
I am also in favor of a dedicated thread 👍 

Since the update is so irregular, it would probably be more efficient to tell the entities when an update is available instead of having Home Assistant ask the entities every 15 seconds to check with the hub if an update is available.

To do that:

 - All entities disable polling (return `False` from `should_poll` property on the entity)
 - All entities call `hass.helpers.dispatcher.dispatcher_connect('mychevy', self.async_update_ha_state)` to listen for updates. (or `async_dispatcher_connect` if in async context
 - Thread will schedule itself. When an update succeeds, call `hass.helpers.dispatcher.dispatcher_send('mychevy')`

Reply:
(the [dispatcher](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/dispatcher.py) is a mini event bus that is for intra-integration communication)